### DISCUSSION TITLE
Communicate with Mindstorms Inventor via USB

### DISCUSSION THREAD

**Discussion by ChristophGaukel** [pybricks/support#431] [Q&A]
Hello,
when I try to communicate from my Linux system with my hub via USB, I'm able to establish a connection with python:

```python
import usb.core

ID_VENDOR = 0x0694
ID_PRODUCT = 0x0010

device = usb.core.find(
    idVendor=ID_VENDOR,
    idProduct=ID_PRODUCT
)
if device is None:
    raise RuntimeError("No Lego mindstorms hub found")

id_serial = usb.util.get_string(device, device.iSerialNumber)
print('Serial ID:', id_serial)
```

This works and prints the serial ID. Now I like to read the sensor data, the battery state etc. Do you have any examples of successful requests and responses? By reading https://lego.github.io/lego-ble-wireless-protocol-docs/index.html I did not get it running.

**Comment by laurensvalk** [comment]
I'm assuming you are running Pybricks on the hub?

In Pybricks, USB is not used for communication. Everything works via Bluetooth (BLE). The hub also doesn't constantly broadcast sensor data using the LEGO wireless protocol.

Although we recommend that people run code via code.pybricks.com, there is also a library called Pybricksdev that you can use to communicate with the hub. It depends a bit on what you are planning to do. :)

**Reply by ChinthaniSugandhika** [reply]
Hi,

I am new to Lego Mindstorms Inventor Hub and I am going to use pybricks. I would like to know the exact communication protocol used to communicate between sensors and the hub. Is it I2C? or LEGO wireless protocol?

**Reply by dlech** [reply]
Hi Chinthani, this sounds like a great question for a [new discussion](https://github.com/pybricks/support/discussions/new) (:arrow_left: click the link to start a new discussion).

**Reply by ChinthaniSugandhika** [reply]
Sure :)

**Comment by ChristophGaukel** [comment]
Thanks for your reply. Seems as if it needs some more information from my side.
My first attempt is to understand the device's capability. E.g. LEGO's apps show the current gyro data, if connected via USB or via Bluetooth and I expect, that the whole functionality of the device is available via Bluetooth requests or USB requests (as the EV3 supports direct commands). I asked about USB but I'm interested in both protocols, USB and Bluetooth.
Maybe, this mode of operation (running the program on my computer and communicating with the hub) is not the focus of Pybricks but it allows to embed the hub into any set of tools, e.g. visualization of sensor data.

**Reply by laurensvalk** [reply]
If the first goal is to understand the device capabilities, I'd recommend experimenting at beta.pybricks.com, which already includes quite a bit of documentation for that hub.

But you can certainly make it work for your use case as well. You can run a script on the hub that prints any desired sensor data, and read it on your PC. Pybricksdev is one way to do this. It is both a command line utility and a Python library that you can integrate in your own tools.

This has the benefit that you can control exactly what information gets sent, and how often. You can even do a bit of pre-processing as needed, like computing an angle of interest so that you don't need to send 6 or more IMU data values all the time. 

If you'd rather do this without running any code on the hub itself, you may want to use the original firmware instead.
